include/d/d_vibration.h

```diff
@@ -1,45 +1,73 @@
 #ifndef D_D_VIBRATION_H
 #define D_D_VIBRATION_H
 
+#include "dolphin/types.h"
 #include "SSystem/SComponent/c_xyz.h"
 
 class dVibration_c {
 public:
-    /* 0x00 */ int field_0x0;
-    /* 0x04 */ int field_0x4;
-    /* 0x08 */ s32 field_0x8;
-    /* 0x0C */ int field_0xc;
-    /* 0x10 */ int field_0x10;
-    /* 0x14 */ cXyz field_0x14;
-    /* 0x20 */ s32 field_0x20;
-    /* 0x24 */ int field_0x24;
-    /* 0x28 */ u32 field_0x28;
-    /* 0x2C */ s32 field_0x2c;
-    /* 0x30 */ int field_0x30;
-    /* 0x34 */ int field_0x34;
-    /* 0x38 */ cXyz field_0x38;
-    /* 0x44 */ s32 field_0x44;
-    /* 0x48 */ int field_0x48;
-    /* 0x4C */ int field_0x4c;
-    /* 0x50 */ s32 field_0x50;
-    /* 0x54 */ s32 field_0x54;
-    /* 0x58 */ s32 field_0x58;
-    /* 0x5C */ s32 field_0x5c;
-    /* 0x60 */ int field_0x60;
-    /* 0x64 */ u32 field_0x64;
-    /* 0x68 */ u32 field_0x68;
-    /* 0x6C */ s32 field_0x6c;
-    /* 0x70 */ s32 field_0x70;
-    /* 0x74 */ s32 field_0x74;
-    /* 0x78 */ s32 field_0x78;
-    /* 0x7C */ int field_0x7c;
+
+    struct vib_pattern {
+        u16 rounds;  /* Number of random bits enabled, used by quakes */
+        u16 length;  /* length of the pattern */
+        s32 pattern; /* pattern bits (interpreted as bitstring) */
+    };
+
+    enum { /* mRumbleState values */
+        RUMBLE_STATE_PAUSED  = -1,
+        RUMBLE_STATE_WAITING =  0,
+        RUMBLE_STATE_RUNNING =  1,
+    };
+    enum { /* Used as flags */
+        RUMBLE_SHOCK = 0x1,
+        RUMBLE_QUAKE = 0x2,
+    };
+
+    /*
+    These weak functions in the debug maps point to an anonymous struct from a precompiled header ?
+    dVibration_c::@class$3553dolzel_pch::@class$3553dolzel_pch()
+    dVibration_c::@class$3553dolzel_pch::@class$3554dolzel_pch::@class$3554dolzel_pch()
+    dVibration_c::@class$3553dolzel_pch::@class$3554dolzel_pch::~@class$3554dolzel_pch()
+    dVibration_c::@class$3553dolzel_pch::~@class$3553dolzel_pch()
+    */
```

Comment:
these anonymous structs seem to be present in TP debug with only minor differences to WW retail, they seem pretty similar to your camera_rumble/motor_rumble except I think they're nested structs inside of each other or something. not really sure exactly what "dolzel.pch" would have been supposed to be though

Reply:
for TP debug the cXyz seems to be at offset 0x18 in the inner struct, and the outer struct contains two instances of the inner struct, rather than one.
![image](https://github.com/user-attachments/assets/ccb0054e-1686-4295-87bf-7a9c1a8ebdd9)
![image](https://github.com/user-attachments/assets/169f9b92-0a1e-4665-9848-a16afc8c9418)

it looks like TP decomp has it like this:
https://github.com/zeldaret/tp/blob/6bf2f3cc5578f3e3108df8337a462ee4c618915a/include/d/d_vibration.h#L27-L40

so maybe something like this:
```cpp
    struct {
        struct {
            /* 0x00 */ int  mPatternIdx;
            /* 0x04 */ u32  mPattern;
            /* 0x08 */ s32  mLength;
            /* 0x0C */ s32  mRounds;
            /* 0x10 */ s32  mFlags;
            /* 0x14 */ cXyz mCoord;
            /* 0x20 */ s32  mCurrentFrame;
        } mShock, mQuake;
    } /* 0x00 */ mCamera;
```

this seems to get the constructors and destructors very close to what debug shows.

and then for consistency, motor was probably like this (even though it doesn't show in debug because there's no cXyz to construct/deconstruct):
```cpp
    struct {
        struct {
          /* 0x00 */ int mPatternIdx;
          /* 0x04 */ u32 mPattern;
          /* 0x08 */ s32 mLength;
          /* 0x0C */ s32 mRounds;
          /* 0x10 */ s32 mCurrentFrame;
          /* 0x14 */ s32 mStopFrame; /* different from length for looping rumble */
        } mShock, mQuake;
    } /* 0x48 */ mMotor;
```

Reply:
Makes a lot more sense, thanks for finding the reference! I've updated accordingly